File: 07pairSum.js

// using a hashmap (object)

// n = length of array
// time:  O(n)
// space: O(n)

const pairSum = (numbers, targetSum) => {
  const previousNums = {};
  for (let i = 0; i < numbers.length; i += 1) {
    const num = numbers[i];
    const complement = targetSum - num;
    if (complement in previousNums) return [i, previousNums[complement]];
    
    previousNums[num] = i;
  }
};


// ------------------------------------------ //

const pairSum = (numbers, targetSum) => {
  // todo
  let nums = {};
  
  for (let idx = 0; idx < numbers.length; idx++) {
    let num = numbers[idx];
    let needed = targetSum - num;
    if (needed in nums) return [nums[needed], idx];
    nums[num] = idx;
  }

};

File: Python/07pairSum.py

# n = length of numbers list
# Time: O(n)
# Space: O(n)

# using a hashmap(dictionary)


def pair_sum(numbers, target_sum):
  previous_numbers = {}

  for index, num in enumerate(numbers):
    complement = target_sum - num

    if complement in previous_numbers:
      return (index, previous_numbers[complement])

    previous_numbers[num] = index


